lua/session_manager/utils.lua

```diff
@@ -196,7 +196,26 @@ end
 function utils.shorten_path(filename)
   -- Shorten path if length exceeds defined max_path_length.
   if config.max_path_length > 0 and #filename > config.max_path_length then
-    return Path:new(filename):shorten()
+    -- Index to exclude from shortening, -1 means last
+    local excludes = { -1 }
+
+    -- Final shortened path to return, return original filename if cannot shorten
+    local shortened = filename
+
+    -- Cache the shortest version
+    local cached = Path:new(filename):shorten(1, excludes)
+
+    -- Gradually increase the tailing excludes
+    while string.len(cached) < config.max_path_length do
+      -- Store the cached path
+      shortened = cached
+
+      -- Try to cache new shortened path with more excludes
+      excludes[#excludes + 1] = excludes[#excludes] - 1
+      cached = Path:new(filename):shorten(1, excludes)
+    end
+
+    return shortened
```

Comment:
Will something like this work? Can't test currently, but the idea is to avoid creating a new Plenary.Path every time and just reuse the previous one.
```suggestion
    local shortened = Path:new(filename):shorten(1, excludes)

    -- Gradually increase the tailing excludes
    while #shortened.filename < config.max_path_length do
      -- Try to short more with more excludes
      excludes[#excludes + 1] = excludes[#excludes] - 1
      shortened = shortened:shorten(1, excludes)
    end
    return shortened.filename
```

Reply:
If I understand correctly, it doesn't seem to work unless we reimplement the functions from plenary.path.

An example for this is `/dir1/dir2/dir3/filename.ext` becomes `/d/d/d/filename.ext` with `excludes = {  -1  }`. With the 2nd iteration where `excludes = { -1, -2 }`, `/d/d/d/filename.ext` is still `/d/d/d/filename.ext` because the text `dir3` is already shortened as `d`.

Please feel free to correct me as I'm not fluent in lua.

Reply:
You are right! I get it wrong for the first time.